subprojects/base-services/src/main/java/org/gradle/internal/classanalysis/AsmConstants.java

```diff
@@ -19,5 +19,5 @@
 import org.objectweb.asm.Opcodes;
 
 public class AsmConstants {
-    public static final int ASM_LEVEL = Opcodes.ASM7;
+    public static final int ASM_LEVEL = Opcodes.ASM8;
```

Comment:
We are using asm library version `9.1` and have `Opcodes.ASM9` and `Opcodes.ASM10_EXPERIMENTAL` available as well.
However, using `ASM9` globally here, breaks Kotlin compilation as Kotlin compiler uses an older asm version.
Perhaps we could separate which asm levels we use in different places.

For, example `ASM9` is required to recognize Java sealed classes. Sealed classes are available since Java 15 in feature preview mode. Incremental compilation will now break if sealed classes are used.

For now, this change only fixes support for Java records that are out of preview mode and available in Java 16.